sys/include/trickle.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Trickle constants and prototypes
+ *
+ * Copyright (C) 2013, 2014  INRIA.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup sys_trickle Trickle Timer
+ * @ingroup sys
+ * @{
+ */
+
+/**
+ * @file    trickle.h
+ * @brief   Implementation of a generic Trickle Algorithm (RFC 6206)
+ *
+ * @author  Eric Engel <eric.engel@fu-berlin.de>
+ * @author  Cenk Gündoğan <cnkgndgn@gmail.com>
+ */
+
+#ifndef _TRICKLE_H
+#define _TRICKLE_H
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+#include "vtimer.h"
+#include "thread.h"
+
+/** @brief a generic callback function with arguments that is called by trickle periodically */
+typedef struct {
+    void (*func)(void *);       /**< a generic callback function pointer */
+    void *args;                 /**< a generic parameter for the callback function pointer */
+} trickle_callback_t;
+
+/** @brief all state variables for a trickle timer */
+typedef struct {
+    uint8_t k;                      /**< redundancy constant */
+    uint32_t Imin;                  /**< minimum interval size */
+    uint8_t Imax;                   /**< maximum interval size, described as a number of doublings */
+    uint32_t I;                     /**< current interval size */
+    uint32_t t;                     /**< time within the current interval */
+    uint16_t c;                     /**< counter */
+    kernel_pid_t pid;               /**< pid of trickles target thread */
+    trickle_callback_t callback;    /**< the callback function and parameter that trickle is calling after each interval */
+    uint16_t interval_msg_type;     /**< the msg_t.type that trickle should use after an interval */
+    timex_t msg_interval_time;      /**< interval represented as timex_t */
+    vtimer_t msg_interval_timer;    /**< vtimer to send a msg_t to the target thread for a new interval */
+    uint16_t callback_msg_type;     /**< the msg_t.type that trickle should use after a callback */
+    timex_t msg_callback_time;      /**< callback interval represented as timex_t */
+    vtimer_t msg_callback_timer;    /**< vtimer to send a msg_t to the target thread for a callback */
+} trickle_t;
+
+/**
+ * @brief resets the trickle timer
+ *
+ * @param[in] trickle   the trickle timer
+ */
+void reset_trickletimer(trickle_t *trickle);
```

Comment:
Public functions should be prefixed with their module name (e.g. `trickle_reset_timer()).